Update ebpf documentation (#11440)
netdata/netdata: collectors/ebpf.plugin/README.md

### System overview

Not all charts within the System Overview menu are enabled by default, because they add around 100ns overhead for each 
function call, this number is small for a human perspective, but the functions are called many times creating an impact 
on host.
See the [configuration](#configuration) section for details about how to enable them.

#### Swap

Inside the swap submenu the eBPF plugin creates the chart `swapcalls`; this chart is displaying when processes are calling
functions `swap_readpage` and `swap_writepage`, which are functions responsible for doing IO in swap memory.

### Memory

In the memory submenu the eBPF plugin creates two submenus **page cache** and **synchronization** with the following
organization:

* Page Cache adds 
  * Page cache ratio
  * Dirty pages
  * Page cache hits
  * Page cache misses
* Synchronization
  * File sync
  * Memory map sync
  * File system sync
  * File range sync

#### Page cache

Page cache is adding `kprobes` for the kernel functions `add_to_page_cache_lru`, `mark_page_accessed`,
`account_page_dirtied`, and `mark_buffer_dirty`. Netdata monitors page cache access giving system-level and 
application-level metrics.

> #### Page cache ratio

Reviewer: "Page cache" is a submenu according to the Memory bulleted list, but "Page cache", "Page cache ratio", and "Dirty pages" have the same heading level (number of `#`).

It is not the only place you have that problem, please go through the "Charts" section and fix where is needed.

If i understand you right, you want to use:
- `###` top level menu section
- `####` sub-menu section
- `#####` a chart (title?)

Author: All right, I moved the data from `Page cache` to submenu and I removed the menu, because they were already explained in `Mem`. 

I removed `Synchronization` without to move content, because the `syscalls` monitored are already cited in chart description.

functions `swap_readpage` and `swap_writepage`, functions responsible for doing IO in swap memory.

### Memory

In memory submenu eBPF plugin creates two submenus **page cache** and **synchronization**.

#### Page cache

This first submenu brings four charts that demonstrate statistics about page cache access.

Page cache is adding `kprobes` for the kernel functions `add_to_page_cache_lru`, `mark_page_accessed`,
`account_page_dirtied`, and `mark_buffer_dirty`. Netdata monitors page cache access giving an overall overview and also
per application access.

##### Page cache ratio

The chart `cachestat_ratio` shows how processes are accessing page cache. In a normal scenario, we expect values around 
 100%, which means that the majority of the work on the machine is processed in memory.

##### Dirty pages

On `cachestat_dirties` netdata demonstrates the number of pages that were modified. This chart reflects calls to function
`mark_buffer_dirty`.

##### Page cache hits

Netdata calls hits when an accessed page cache was not modified and we do not count pages that were added recently. This

> chart uses data from the four functions we are monitoring, mentioned above.

Author: You understood correctly, I am using your suggestion.

## Charts

The eBPF collector creates an **eBPF** menu in the Agent's dashboard along with two sub-menus: **Socket**, and

> **Process**. The plugin also creates charts on different menus, like System Overview, Memory, Disks, Filesystem, 

Reviewer 2: should menu names be enclosed in `backticks` (``)? we do that for chart names below

Author: A good point, I thought the bold around would be enough, @hugovalente-pm can you clarify this for us?

Reviewer 2: > System Overview, Memory, Disks, Filesystem, Mount Points and Applications

Is neither bold nor inline code (with backticks)

Reviewer 3: was trying to find other examples and could only find this one https://learn.netdata.cloud/docs/agent/collectors/apps.plugin where bold is used for **Applications**, **Users** and **User Groups**, but this was the only I could find not sure if there are other examples.

imo bold works ok as long as we do on all, like the place @underhood called, but this is just personal opinion

Author: @underhood this is the unique change I did not bring, I will wait for your answer to proceed.

Reviewer 2: @thiagoftsm up to you then, wont block on this

Author: Thank you! I will move forward with @hugovalente-pm 's suggestion, but we will have more PRs related to these group of charts, so we will have opportunity to change them again.

Mount Points and Applications. All the charts in this section update every second.
The collector stores the actual value inside of its process, but charts only show the difference between the values
collected in the previous and current seconds.

### System Overview

> All charts on system overview are not enabled by default, because they would consume a lot of resources.

Reviewer 2: I think this sentence is not clear and could be understood (at least by me) in 2 ways:
- All charts are disabled by default
- Not all of the charts are enabled by default (some are)

Reviewer: "...consume a lot of resources" sounds scary, i would never enable all the charts 😄

Author: ping @hugovalente-pm !

Author: @ilyam8 when I address the latency areas, one of the features I will bring it will be to extend the time we are reading data from kernel ring. These eBPF programs are on top of heavy operations, this is the motive we brought this warning.

Reviewer 2: > "...consume a lot of resources" sounds scary, i would never enable all the charts

All your CPUz belong to us!

Reviewer 3: @thiagoftsm @ilyam8 from what I understood these configurations should only be enabled if someone is trying to troubleshoot and needs to have full visibility into the date we collect with eBPF, right?

I believe the callout is needed, what we could add is an highlight to show in which case a user could want to enabled this and what would be the benefits of doing it

Reviewer: I agree that warning is needed. My point is that being more specific is better - what resources (cpu, mem, etc), how many resources.

Author: I added some additional description trying to let it more soft, when the latest version of `libbpf` is merged during this week, we will be able to reduce the overhead, because new features will be available for us, but these features depend of kernel compilation, so we need to keep some warnings like this.

Reviewer 3: @thiagoftsm, as @ilyam8  suggedted, can we name and, if possible, measure what are the impacts in terms of resources and a value of % of increase in usage?

Author: @hugovalente-pm it is possible, and I will measure to bring in the next PR, for this PR I would prefer to add this missing documentation for we have the links, in the next PR, that I expect to bring next days, I will add a table with these values.

I accepted @underhood suggestion.

functions `swap_readpage` and `swap_writepage`, functions responsible for doing IO in swap memory.

### Memory

In memory submenu eBPF plugin creates two submenus **page cache** and **synchronization**.

#### Page cache

This first submenu brings four charts that demonstrate statistics about page cache access.

Page cache is adding `kprobes` for the kernel functions `add_to_page_cache_lru`, `mark_page_accessed`,
`account_page_dirtied`, and `mark_buffer_dirty`. Netdata monitors page cache access giving an overall overview and also
per application access.

##### Page cache ratio

The chart `cachestat_ratio` shows how processes are accessing page cache. In a normal scenario, we expect values around 
 100%, which means that the majority of the work on the machine is processed in memory.

##### Dirty pages

On `cachestat_dirties` netdata demonstrates the number of pages that were modified. This chart reflects calls to function
`mark_buffer_dirty`.

##### Page cache hits

Netdata calls hits when an accessed page cache was not modified and we do not count pages that were added recently. This

> chart uses data from the four functions we are monitoring, mentioned above,.

Reviewer 3: ```suggestion
chart uses data from the four functions we are monitoring, mentioned above.
```

##### Page cache misses

A page cache miss means that a page was not inside memory when the process tried to access it. This chart shows the result
of the difference for calls between functions `add_to_page_cache_lru` and `account_page_dirtied`.

> #### Syncrhonization

Reviewer 3: ```suggestion
#### Synchronization
```

> The **synchronization** submenu Netdata monitors calls

Reviewer 2: Either `The **synchronization** submenu OF Netdata monitors calls...` or `IN the **synchronization** submenu, Netdata monitors...`

Reviewer 2: also `calls for different syscalls` sounds confusing to me. Probably would put it as `monitors usage of various kernel syscalls`

for different syscalls.

##### File sync

> This chart shows calls to synchronization methods, fsync() and fdatasync(), to transfer all modified page caches for the 

Reviewer 2: ```suggestion
This chart shows calls to synchronization methods, `fsync()` and `fdatasync()`, to transfer all modified page caches for the 
```

files on disk devices. These calls block until the disk reports that the transfer has been completed.
They flush data for specific file descriptors.

> ##### Memory map sync

Reviewer 2: Should probably unify capitalization of captions? We have
- Memory map sync (Mms)
and below it
- File System sync
e.g. every caption seems to use different rules

Author: Thank you to call attention for this, I changed and now only the first letter of the sentence is upper case.

The chart shows calls to `msync(2)` syscalls. This syscall flushes changes of a file that was mapped into memory using
`mmap(2)`.

##### File System sync

This chart monitors calls demonstrating commits from filesystem caches to disk.

##### File Range sync

This chart shows calls to `sync_file_range(2)` which synchronizes file segments with disk. This is the most dangerous 
syscall to synchronize data according to its manual.

### Disk

The eBPF plugin also shows a chart on Disk section when `disk` thread is enabled. This will create the
chart `disk_latency_io` for each disk on the host. These charts use [tracepoints](https://www.kernel.org/doc/html/latest/trace/tracepoints.html)
`block:block_rq_issue` and `block:block_rq_complete` to measure the needed time for an IO event to be completed.

### Filesystem

> This group has two charts demonstrating how applications interacts with the Linux kernel to open and close file descriptors.

Reviewer 3: ```suggestion
This group has two charts demonstrating how applications interact with the Linux kernel to open and close file descriptors.
```

collected in the previous and current seconds.

### System overview

Not all charts within the System Overview menu are enabled by default, because they add around 100ns overhead for each 
function call, this number is small for a human perspective, but the functions are called many times creating an impact 
on host.
See the [configuration](#configuration) section for details about how to enable them.

#### Swap

Inside the swap submenu the eBPF plugin creates the chart `swapcalls`; this chart is displaying when processes are calling
functions `swap_readpage` and `swap_writepage`, which are functions responsible for doing IO in swap memory.

### Memory

In the memory submenu the eBPF plugin creates two submenus **page cache** and **synchronization** with the following
organization:

* Page Cache
  * Page cache ratio
  * Dirty pages
  * Page cache hits
  * Page cache misses
* Synchronization
  * File sync
  * Memory map sync
  * File system sync
  * File range sync

#### Page cache ratio

The chart `cachestat_ratio` shows how processes are accessing page cache. In a normal scenario, we expect values around 

>  100%, which means that the majority of the work on the machine is processed in memory. To calculate the ratio Netdata

Reviewer 4: There are many places with whitespaces at the beginning of strings or at the end, like in these lines. Please remove them.

Author: Thanks, I removed.

#### Create

This chart shows the number of calls to `vfs_create`. This function is responsible for creating files.

#### Synchronization

This chart shows the number of calls to `vfs_fsync`. This function is responsible for performing a fsync or fdatasync 

> on a file. You can see more details on Synchronization

Reviewer 3: Would be good to link this to the line 90 (#syncrhonization) but I'm not sure how this will work, since we have two sub-sections with Synchronization name...

Author: I agree, let us merge this code like it is and when I bring more updates with other PRs, we create the links.

#### Create

This chart shows the number of calls to `vfs_create`. This function is responsible for creating files.

#### Synchronization

This chart shows the number of calls to `vfs_fsync`. This function is responsible for calling `fsync(2)` or 

> `fdatasync(2)` on a file. You can see more details on in the Synchronization section.

Reviewer 4: ```suggestion
`fdatasync(2)` on a file. You can see more details in the Synchronization section.
```

Reviewer 4: Hmm, I didn't ask you to change the first `on`. I asked to remove the second one.

#### Open

This chart shows the number of calls to `vfs_open`. This function is responsible for opening files.

### eBPF

#### Process thread

Internally, the Linux kernel treats both processes and threads as `tasks`. To create a thread, the kernel offers a few

> system calls: `fork(2)`, `vfork(2)` and `clone(2)`. In turn, each of these system calls uses either the function

Reviewer 4: ```suggestion
system calls: `fork(2)`, `vfork(2)`, and `clone(2)`. In turn, each of these system calls uses either the function
```

#### Task error

The functions responsible for ending tasks do not return values, so this chart contains information about failures on 
process and thread creation only.

#### TCP functions

This chart demonstrates calls to functions `tcp_sendmsg`, `tcp_cleanup_rbuf` and `tcp_close`, these functions are
used to send, and receive data and to close connections when `TCP` protocol is used.

#### TCP bandwidth

Like the previous section, this chart also monitors `tcp_sendmsg` and `tcp_cleanup_rbuf` but, instead of showing number of calls,
it demonstrates the number of bytes sent and received.

#### TCP retransmit

This chart demonstrates calls to functions `tcp_retransmit` that is responsible to execute the TCP retransmission when the
receiver did not return the package during the expected time.

#### UDP functions

> This chart demonstrates calls to  functions `udp_sendmsg`, and `udp_recvmsg` responsible to send and receive data

Reviewer 5: ```diff
-`udp_sendmsg`, and `udp_recvmsg` responsible to send and receive
+`udp_sendmsg` and `udp_recvmsg`, which are responsible for sending & receiving
```

> for connections when `UDP` protocol is used.

Reviewer 5: ```diff
-when `UDP` protocol
+when the `UDP` protocol
```

#### UDP bandwidth

> Like previous section, this chart also monitors `tcp_sendmsg` and `tcp_cleanup_rbuf`but,  instead of showing the number of calls,

Reviewer 5: ```diff
-Like previous section,
+Like the previous section,
```

```diff
-`tcp_sendmsg` and `tcp_cleanup_rbuf`but,  instead of
+`tcp_sendmsg` and `tcp_cleanup_rbuf`, but instead of
```

#### Task error

The functions responsible for ending tasks do not return values, so this chart contains information about failures on 
process and thread creation only.

#### TCP functions

> This chart demonstrates calls to functions `tcp_sendmsg`, `tcp_cleanup_rbuf` and `tcp_close`; these functions are

Reviewer 4: ```suggestion
This chart demonstrates calls to functions `tcp_sendmsg`, `tcp_cleanup_rbuf`, and `tcp_close`; these functions are
```

Reviewer 5: Actually a comma before the "and" in a list is not necessary. At least that's what my 9th grade English teacher used to insist 😄

Reviewer 5: Alright well seems like it's "optional": https://www.grammarly.com/blog/comma-before-and/

Reviewer 4: https://learn.netdata.cloud/contribute/style-guide#oxford-comma

Author: @UmanShahzad I also heard this a lot not only in English classes, but for Latin languages too. :smile:

used to send & receive data and to close connections when `TCP` protocol is used.

#### TCP bandwidth

> Like the previous section, this chart also monitors `tcp_sendmsg` and `tcp_cleanup_rbuf` but, instead of showing the 

Reviewer 4: ```suggestion
Like the previous chart, this one also monitors `tcp_sendmsg` and `tcp_cleanup_rbuf` but, instead of showing the 
```
BTW, why there is the `close` dimension in this chart? Shouldn't it be deleted?

Reviewer 2: Also comma here should be before not after "but"

Reviewer 5: @vlvkobal I believe @underhood is right.

Author: @vlvkobal you are right, I will remove the dimension in another PR.

@underhood and @UmanShahzad  I had a long talk with @hugovalente-pm about this, and I am glad I am not the unique with this opinion, I am moving the comma.

number of calls, it demonstrates the number of bytes sent and received.

#### TCP retransmit

This chart demonstrates calls to functions `tcp_retransmit` that is responsible for executing TCP retransmission when the
receiver did not return the packet during the expected time.

#### UDP functions

This chart demonstrates calls to functions `udp_sendmsg` and `udp_recvmsg`,  which are responsible for sending & 
 receiving data for connections when the `UDP` protocol is used.

#### UDP bandwidth

> Like the previous section, this chart also monitors `tcp_sendmsg` and `tcp_cleanup_rbuf`, but instead of showing the 

Reviewer 4: ```suggestion
Like the previous chart, this one also monitors `udp_sendmsg` and `udp_recvmsg`, but instead of showing the 
```
